phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java

```diff
@@ -3416,6 +3437,167 @@ private MetaDataMutationResult mutateColumn(
+     * @throws Throwable
+     */
+    private void invalidateServerMetadataCache(byte[] tenantId, byte[]schemaName,
+            byte[] tableOrViewName) throws Throwable {
+        Configuration conf = env.getConfiguration();
+        String value = conf.get(REGIONSERVER_COPROCESSOR_CONF_KEY);
+        if (value == null
+                || !value.contains(PhoenixRegionServerEndpoint.class.getName())) {
+            // PhoenixRegionServerEndpoint is not loaded. We don't have to invalidate the cache.
+            LOGGER.info("Skip invalidating server metadata cache for tenantID: {},"
+                            + " schema name: {}, table Name: {} since PhoenixRegionServerEndpoint"
+                            + " is not loaded", Bytes.toString(tenantId),
+                    Bytes.toString(schemaName), Bytes.toString(tableOrViewName));
+            return;
+        }
+        Properties properties = new Properties();
+        // Skip checking of system table existence since the system tables should have created
+        // by now.
+        properties.setProperty(SKIP_SYSTEM_TABLES_EXISTENCE_CHECK, "true");
+        try (PhoenixConnection connection = QueryUtil.getConnectionOnServer(properties,
+                env.getConfiguration()).unwrap(PhoenixConnection.class);
+             Admin admin = connection.getQueryServices().getAdmin()) {
+            // This will incur an extra RPC to the master. This RPC is required since we want to
+            // get current list of regionservers.
+            Collection<ServerName> serverNames = admin.getRegionServers(true);
+            invalidateServerMetadataCacheWithRetries(admin, serverNames, tenantId, schemaName,
+                    tableOrViewName, false);
+        }
+    }
+
+    /**
+     * Invalidate metadata cache on all regionservers with retries for the given tenantID
+     * and tableName with retries. We retry once before failing the operation.
+     *
+     * @param admin
+     * @param serverNames
+     * @param tenantId
+     * @param schemaName
+     * @param tableOrViewName
+     * @param retryAttempt
+     * @throws Throwable
+     */
+    private void invalidateServerMetadataCacheWithRetries(Admin admin,
+            Collection<ServerName> serverNames, byte[] tenantId, byte[] schemaName,
+            byte[] tableOrViewName, boolean retryAttempt) throws Throwable {
+        String fullTableName = SchemaUtil.getTableName(schemaName, tableOrViewName);
+        String tenantIDStr = Bytes.toString(tenantId);
+        LOGGER.info("Invalidating metadata cache for tenantID: {}, tableName: {} for"
+                        + " region servers: {}, isRetryAttempt: {}", tenantIDStr, fullTableName,
+                serverNames, retryAttempt);
+        RegionServerEndpointProtos.InvalidateServerMetadataCacheRequest request =
+                getRequest(tenantId, schemaName, tableOrViewName);
+        // TODO Do I need my own executor or can I re-use QueryServices#Executor
```

Comment:
What is the threadpool size of the default executor ? Maybe it is better to create a new executor so that you can control its size

Reply:
> What is the threadpool size of the default executor ? 

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#ForkJoinPool--
Currently default is ForkJoinPool which uses the number of threads as many as available processors.

Reply:
>  Maybe it is better to create a new executor so that you can control its size

@tkhurana  Do you think currently we can keep it as it is and use custom threadpool if needed?

Reply:
Sounds good @shahrs87